jetbrains-core/src/software/aws/toolkits/jetbrains/core/stack/TableView.kt

```diff
@@ -0,0 +1,104 @@
+import com.intellij.ui.components.JBScrollPane
+import com.intellij.ui.table.JBTable
+import software.amazon.awssdk.services.cloudformation.model.ResourceStatus
+import software.amazon.awssdk.services.cloudformation.model.StackEvent
+import java.awt.FlowLayout
+import javax.swing.BoxLayout
+import javax.swing.JButton
+import javax.swing.JComponent
+import javax.swing.JLabel
+import javax.swing.JPanel
+import javax.swing.JTable
+import javax.swing.SwingConstants
+import javax.swing.table.DefaultTableCellRenderer
+import javax.swing.table.DefaultTableModel
+
+/**
+ * Events table
+ */
+internal interface TableView {
+    /**
+     * @param events List<StackEvent> Events in reverse chronological order.
+     * All events are inserted in the top of the table.
+     */
+    fun insertEvents(events: List<StackEvent>)
+
+    val component: JComponent
+}
+
+private enum class Fields(val readableName: String, val getData: (StackEvent) -> Any) {
+    TIME("Time", { e -> e.timestamp() }),
+    STATUS("Status", { e -> e.resourceStatus() }),
+    TYPE("Type", { e -> e.resourceType() }),
+    LOGICAL_ID("Logical ID", { e -> e.logicalResourceId() }),
+    PHYSICAL_ID("Physical ID", { e -> e.physicalResourceId() });
+
+    override fun toString() = readableName
+}
+
+private class StatusCellRenderer : DefaultTableCellRenderer() {
+    override fun getTableCellRendererComponent(table: JTable, value: Any, isSelected: Boolean, hasFocus: Boolean, row: Int, column: Int) =
+        (value as ResourceStatus).run { JLabel(name, type.icon, SwingConstants.LEFT) }
+}
+
+private class StackTableModel : DefaultTableModel(Fields.values().map(Fields::readableName).toTypedArray(), 0) {
+    override fun isCellEditable(row: Int, column: Int) = false
+}
+
+/**
+ * @property onButton Function1<Boolean, Unit> user clicked button (forward:Bool). Called on EDT
+ */
+internal class TableViewImpl(private val onButton: (Boolean) -> Unit) : TableView, Disposable {
+
+    private val model = StackTableModel()
+    private val table = JBTable(model).apply {
+        autoCreateRowSorter = true
+        autoscrolls = true
+        setShowColumns(true)
+        setPaintBusy(true)
+        getColumn(Fields.STATUS.readableName).cellRenderer = StatusCellRenderer()
+    }
+    override val component: JComponent = JPanel()
+    private val prevButton = createButton(forward = false)
+    private val nextButton = createButton(forward = true)
+
+    init {
+        component.apply {
+            layout = BoxLayout(component, BoxLayout.Y_AXIS)
+            add(JBScrollPane(table))
+            add(JPanel(FlowLayout(FlowLayout.RIGHT)).apply {
+                add(prevButton)
+                add(nextButton)
+            })
+        }
+    }
+
+    fun enableButtons(prev: Boolean, next: Boolean) {
+        prevButton.isEnabled = prev
+        nextButton.isEnabled = next
+    }
+
+    private fun createButton(forward: Boolean) =
+        JButton(if (forward) ">>" else "<<").apply {
```

Comment:
How does it look using com.intellij.icons.AllIcons.Actions#Forward / Back

Reply:
Done